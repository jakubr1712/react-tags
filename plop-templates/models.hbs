export interface I{{properCase name}}ComponentProps {}