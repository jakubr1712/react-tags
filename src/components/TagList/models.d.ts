export interface ITagListProps {
  tags: CommonTypes.ITag[];
  filterText: string;
  toggleTag: (tag: CommonTypes.ITag) => void;
}