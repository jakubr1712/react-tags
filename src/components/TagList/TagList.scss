.tag-list {
  list-style: none;
  padding: 0;
  margin: 0;

  &__item {
    display: flex;
    width: 100%;
    align-items: center;
    padding: 10px 3px;
    cursor: pointer;

    &:hover {
      background-color: #f5f5f5;
    }

    &--highlighted {
      background-color: #f5f5f5;
    }
  }

  &__name {
    margin-left: 10px;
  }

  &__count {
    right: 20px;
    position: absolute;
    font-weight: 300;
  }
}
