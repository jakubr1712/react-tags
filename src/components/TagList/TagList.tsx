import React, { FC } from "react";
import "./TagList.scss";
import { ITagListProps } from "./models.d";
import TagListItem from "../TagListItem";

const TagList: FC<ITagListProps> = ({ tags, filterText, toggleTag }) => {
  const tagLists = tags
    .filter((tag) => tag.name.toLowerCase().includes(filterText.toLowerCase()))
    .map((tag, index) => (
      <TagListItem tag={tag} index={index} toggleTag={toggleTag} />
    ));
  return <ul className="tag-list">{tagLists}</ul>;
};

export default TagList;
