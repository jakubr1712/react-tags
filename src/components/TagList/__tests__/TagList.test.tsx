import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import TagList from "../TagList";
import { ITagListProps } from "../models.d";
import { ITagListItemProps } from "../../TagListItem/models.d";

jest.mock("../../TagListItem", () => (props: ITagListItemProps) => (
  <li>{props.tag.name}</li>
));

describe("TagList component", () => {
  const mockToggleTag = jest.fn();

  const defaultProps: ITagListProps = {
    tags: [
      { id: 1, name: "Tag1", count: 10, isActive: false },
      { id: 2, name: "Tag2", count: 20, isActive: true },
      { id: 3, name: "Tag3", count: 30, isActive: false },
    ],
    filterText: "",
    toggleTag: mockToggleTag,
  };

  beforeEach(() => {
    mockToggleTag.mockClear();
  });

  it("renders properly and matches snapshot", () => {
    const { container } = render(<TagList {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the correct number of TagListItem components", () => {
    render(<TagList {...defaultProps} />);
    expect(screen.getByText("Tag1")).toBeInTheDocument();
    expect(screen.getByText("Tag2")).toBeInTheDocument();
    expect(screen.getByText("Tag3")).toBeInTheDocument();
  });

  it("filters tags based on filterText", () => {
    render(<TagList {...defaultProps} filterText="Tag1" />);
    expect(screen.getByText("Tag1")).toBeInTheDocument();
    expect(screen.queryByText("Tag2")).not.toBeInTheDocument();
    expect(screen.queryByText("Tag3")).not.toBeInTheDocument();
  });

  it("passes the correct props to TagListItem", () => {
    render(<TagList {...defaultProps} />);
    const filteredTags = defaultProps.tags.filter((tag) =>
      tag.name.toLowerCase().includes(defaultProps.filterText.toLowerCase())
    );

    filteredTags.forEach((tag, index) => {
      expect(screen.getByText(tag.name)).toBeInTheDocument();
    });
  });

  it("renders the correct tag names", () => {
    render(<TagList {...defaultProps} />);
    defaultProps.tags.forEach((tag) => {
      expect(screen.getByText(tag.name)).toBeInTheDocument();
    });
  });
});
