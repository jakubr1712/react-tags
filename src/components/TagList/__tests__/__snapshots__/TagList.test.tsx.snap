// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagList component renders properly and matches snapshot 1`] = `
<div>
  <ul
    class="tag-list"
  >
    <li>
      Tag1
    </li>
    <li>
      Tag2
    </li>
    <li>
      Tag3
    </li>
  </ul>
</div>
`;
